subprojects/core/src/main/java/org/gradle/internal/classpath/ClasspathBuilder.java

```diff
@@ -125,5 +131,26 @@ private ZipEntry newZipEntryWithFixedTime(String name) {
             entry.setTime(ZipCopyAction.CONSTANT_TIME_FOR_ZIP_ENTRIES);
             return entry;
         }
+
+        private void configureCompression(ZipEntry entry, CompressionMethod compressionMethod, byte[] contents) {
+            if (shouldCompress(compressionMethod)) {
+                entry.setMethod(ZipEntry.DEFLATED);
+            } else {
+                entry.setMethod(ZipEntry.STORED);
+                // A stored ZipEntry requires setting size and CRC32 upfront.
+                // See https://stackoverflow.com/q/1206970.
+                entry.setSize(contents.length);
+                entry.setCompressedSize(contents.length);
+                entry.setCrc(Hashing.crc32().hashBytes(contents).padToLong());
+            }
+        }
+
+        private static boolean shouldCompress(CompressionMethod compressionMethod) {
+            // Stored files may be used for memory mapping, so it is important to store them uncompressed.
+            // All other files are fine being compressed to reduce on-disk size.
+            // It isn't clear if storing them uncompressed too would bring a performance benefit,
+            // as reading less from the disk may save more time than spent unpacking.
```

Comment:
This also depends on what resource is more available (CPU or I/O), but in general unzipping tends to be a lot faster than reading stuff from disk uncompressed.

Reply:
Noted, though I don't want to put strong words here without a supporting benchmark.